### ISSUE TITLE
Meross MSS425E Power Switches

### ISSUE THREAD

**Issue by RusuPetrut** [theyosh/TerrariumPI#254]
Hi guys ! What a fabulous job you've done here  !!!  I'm a newbie in Terrarium Pi and just play around until i understand it and after i will put this system on my aquarium . I'm excited to join the team !

I have a question/request if you don't mind , i bought before searching and finding Terrarium PI , two Power Switches "Meross MSS425E" , which i find them very handy with my current system(2 pumps , 1 heater, lights , air stone , co2 electrovalve) but i want to integrate them with Terrarium Pi . 

Those power switches are cheap (30E) and you can find them on amazon , here is a link : https://amzn.to/2SFvnTZ

I found that Meross integrates with Google Home and Alexa and somebody already try to integrate them and published a library : https://pypi.org/project/meross-iot/

Could you please provide a feedback about this ? 

Thank you for everything ! 
![whatsapp image 2019-01-26 at 16 08 25](https://user-images.githubusercontent.com/17450475/52520160-3abd7e80-2c6f-11e9-8ed2-97517c1a75a0.jpeg)
![whatsapp image 2019-01-26 at 16 08 21](https://user-images.githubusercontent.com/17450475/52520161-3b561500-2c6f-11e9-8edf-9cef0e10b6c4.jpeg)
![whatsapp image 2019-01-26 at 16 08 22 1](https://user-images.githubusercontent.com/17450475/52520162-3b561500-2c6f-11e9-9de7-8cd3e9b2f932.jpeg)
![Uploading WhatsApp Image 2019-01-26 at 16.08.22.jpeg…]()
![Uploading WhatsApp Image 2019-01-26 at 16.08.24.jpeg…]()

**Comment by theyosh** [comment]
Hi,

I think that it is possible to integrate those power plugs. They act almost as WeMo power plugs. But I do not have the hardware at home, nor will I buy it (for personal reasons). So then I would ask you to do some testing and debugging for me.

If you have a Pi running, please install the python library:

`pip install meross_iot --upgrade`

Then create a `test.py` file with the following contents:
```
import time
import sys
from meross_iot.api import MerossHttpClient

if __name__=='__main__':
    httpHandler = MerossHttpClient(email="YOUR_MEROSS_CLOUD_EMAIL", password="YOUR_PASSWORD")

    # Retrieves the list of supported devices
    print("Listing Devices...")
    devices = httpHandler.list_supported_devices()
    print(devices)

    for counter, device in enumerate(devices):
        print("Playing with device: %d" % counter)
        print(device)
        # Returns most of the info about the power plug
        print("\nGetting system data...")
        data = device.get_sys_data()
        print(data)

        # Reads the historical device consumption
        print("\nReading consumption data...")
        consumption = device.get_power_consumptionX()
        print(consumption)

        # Info about the device
        print("\nGetting device trace...")
        trace = device.get_trace()
        print(trace)
        print("\nGetting device debug...")
        debug = device.get_debug()
        print(debug)

        # Returns the capabilities of this device
        print("\nRetrieving device abilities...")
        abilities = device.get_abilities()
        print(abilities)

        # Returns the current power consumption and voltage from the plug
        # (Note: this is not really realtime, but close enough)
        print("\nReading electricity...")
        electricity = device.get_electricity()
        print(electricity)
```

You need to fill in your login details for the service. If you then run the file with python it should produce some output. I need that output to see what information is available to work with.

So run: `python test.py` and make either a screenshot or copy paste the output here. By copy pasting is formatting needed.

**Comment by RusuPetrut** [comment]
Hi , 

I just run "python test.py" , here is the response from the  Meross MSS425E Power Strip : 

![putty_rklhdahqdj](https://user-images.githubusercontent.com/17450475/52535934-d11b9e00-2d5c-11e9-9305-3f5fcb44f037.png)

I also format the response and you can download from here : [https://we.tl/t-Et3RKnPC4c](url)

Thank you

**Comment by theyosh** [comment]
Hmm, is see that your device is missing some code support. So therefore the above test does give an error. Could you try the following:

Edit the file `test.py` and remove the following lines:

```
# Reads the historical device consumption
print("\nReading consumption data...")
consumption = device.get_power_consumptionX()
print(consumption)
```

and rerun again: `python test.py`. It should produce more output.

**Comment by RusuPetrut** [comment]
Thanks for your quick reply , here is the output : [https://gist.github.com/RusuPetrut/1c327ea14766baa824b4af270ec5af9f](url)

**Comment by theyosh** [comment]
Yes! Ok, now I need some time to format that information correctly and see what I can do with it. It can take a few days before I have it working.

But thanks a lot for the information!

**Comment by theyosh** [comment]
Is it correct that with the last test, all power plugs where off? And do you have 4 power plug connections and 1 for usb?

**Comment by RusuPetrut** [comment]
Yes , all the plugs were off at test point , because before the script modification i just played with toggle to see what channel correspond to each plug , here is the conclusion : 
![paintdotnet_iax9ajisut](https://user-images.githubusercontent.com/17450475/52536768-89e6da80-2d67-11e9-9bde-dcf99db4b686.png)

Only Ch1 , Ch2 , Ch3 are necessary for toggle.

**Comment by theyosh** [comment]
But does the output repeat it then 5 times? Because I can only do an on or off at a device. And toggle all at once is not what you want..

**Comment by RusuPetrut** [comment]
Yes , i repet the output test `test.py` 5 times, each time I toggle one plug with the Meross Aplication. If this is what you mean

**Comment by theyosh** [comment]
No, I had the idea that a device is one plug and therefore the output should be repeated. But as it turns out, a device is just the hole power device with different channels. The used example and test script did not make that clear. So there is no repeat.

But that you see a change from 1 to 0 is good and does confirm that all info is available.

**Comment by theyosh** [comment]
Ok, after some thinking I think I have a working way. But I could not test it. So I am asking you to test this for me. But it needs Python 3, so I hope you have some time to set it up for Python 3:

- `cd TerrariumPI`
- `git pull`
- `git checkout issue/254`
- `sudo ./install.sh 3` ( the value 3 will install everything for using Python 3 including the new Merros IoT module)
- make sure there is no TerrariumPI running. https://github.com/theyosh/TerrariumPI/wiki/FAQ#how-to-stop-terrariumpi
- `sudo ./start.sh 3` ( again the 3 is telling using Python 3)
- **clear browser cache**
- Go to the System settings, and you should see a new part called 'Merros Cloud'. Enter here you login email address and password. Save it
- When the new settings are saved, you have to restart TerrariumPI in the screen session. Do not restart the Pi itself. https://github.com/theyosh/TerrariumPI/wiki/FAQ#how-to-restart-terrariumpi
- When restarted, it should show at least 4 new power switches. If not there should be errors in your screen session. https://github.com/theyosh/TerrariumPI/wiki/FAQ#how-to-debug-terrariumpi

For now this is the easiest way to add support for those power switches. The restart after setting login credentials is needed. But once added and the system has found your powerplugs, there is no restart needed anymore and everything should work.

Also, keep in mind, I do **NOT** encrypt password in the settings. So you Merros cloud password will be stored in plain text!

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52656069-3f4b9680-2efe-11e9-8bf0-d46a13ce2fc8.JPG)

Hi ! This is what i receive after i install everything for using Python 3 `sudo ./install.sh 3`  and the systems reboots

**Comment by theyosh** [comment]
Ok, you found a bug. So I fixed that now. So execute the command in the TerrariumPI folder

`git pull`

And then restart it. It should at least go further then this error

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52661720-52189800-2f0b-11e9-9bcf-b74513f2982d.JPG)

No "Merros Cloud" in system settings .

**Comment by theyosh** [comment]
Did you clear your browser cache?

**Comment by RusuPetrut** [comment]
Yes

**Comment by theyosh** [comment]
And are you sure that you are on the branch 'issue/254' ? Because it should show a second row with title for the meross cloud settings.

**Comment by RusuPetrut** [comment]
![a](https://user-images.githubusercontent.com/17450475/52663012-4f6b7200-2f0e-11e9-9931-40c6115f1374.JPG)
I receive this error when i want to go on branch , do you know wy ?

**Comment by theyosh** [comment]
This is strange. You have done a `git pull` which should provide the latest information. Also this branch.

If I type in `git checkout issue/2` and then a Tab, it should give all the available branches. And there I see this one also. I have no idea why you do not have it

**Comment by theyosh** [comment]
oh crap... I do know why you do not see it... My installation instructions tells you not to download multiple branches.... Crap crap...

So, then it needs some more work. First, move the current TerrariumPI to a temp location or rename it.

Then do the following command:

`git clone --recursive https://github.com/theyosh/TerrariumPI.git`

This will get all history and branches in the TerrariumPI folder. Then you should be able to go to the branch issue/254

**Comment by theyosh** [comment]
But that also means, that the you have to rerun the installer as well. All you actions are done on the master branch. So that means after the new git clone start again at: https://github.com/theyosh/TerrariumPI/issues/254#issuecomment-462525080 at the 3rd dot

**Comment by RusuPetrut** [comment]
I manage to install it , when i try to "submit" the switch i receive this error , the system don't save this setting .

![capture1](https://user-images.githubusercontent.com/17450475/52666116-07504d80-2f16-11e9-89ea-1c4f4eac9a7c.JPG)
![capture](https://user-images.githubusercontent.com/17450475/52666123-0d462e80-2f16-11e9-8fba-6bf608ec0fb3.JPG)

**Comment by theyosh** [comment]
Made again some fixes. I hope that this does not bother you...  that you have to try this a couple of times.

So do again `git pull` and restart. Then you should be able to save the settings. When that succeeds, restart once more to activate the Meross cloud

**Comment by RusuPetrut** [comment]
No problem , glad to help you . The settings are the same from the previous post 
![capture](https://user-images.githubusercontent.com/17450475/52667611-f275b900-2f19-11e9-85a6-f95a19439f1c.JPG)

**Comment by theyosh** [comment]
Cool. So next run: `git pull` and restart.

**Comment by RusuPetrut** [comment]
Yes , the authentication is working . I receive the same error
![capture](https://user-images.githubusercontent.com/17450475/52668179-8300c900-2f1b-11e9-9b96-1471e572361a.JPG)

**Comment by theyosh** [comment]
yesh... I just found another bug.. Again `git pull` and restart :)

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52668423-2e118280-2f1c-11e9-9830-eb831cb61297.JPG)

**Comment by theyosh** [comment]
Are you running this with Python 3? Because that is mandatory. So you should start it with `sudo ./start.sh 3` the 3 value is important here

**Comment by RusuPetrut** [comment]
Yes, always `sudo ./start.sh 3`

**Comment by theyosh** [comment]
Hmm, this is a bit hard. But I made a change, and added some more debug code. So again `git pull` and restart please.

**Comment by theyosh** [comment]
And, you should not have to save the settings. Only restart it, and let it run...

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52668785-2a323000-2f1d-11e9-82f4-2fb0a4f96aab.JPG)

**Comment by theyosh** [comment]
Ok, found a stupid typo. So, `git pull` and restart. And just restart. It looks like you are saving some settings? That should not be needed any more. So, for now, the goal is that it starts without Traceback errors.

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52669170-1804c180-2f1e-11e9-9761-3fa3f9642ecd.JPG)

**Comment by theyosh** [comment]
And yesh, found another typo.. :) `git pull` and restart

**Comment by RusuPetrut** [comment]
Ok , now all power plugs after `sudo ./start.sh 3` just go off 

![capture](https://user-images.githubusercontent.com/17450475/52669666-4d5ddf00-2f1f-11e9-9360-b6db9f4d2fdf.JPG)

**Comment by theyosh** [comment]
Ok , this one is a bit hard.. I do have an idea, but not sure. So I added some extra debug. Lets see what that says. `git pull` and restart please.

**Comment by theyosh** [comment]
just found the bug... make sure you do an extra `git pull`

**Comment by RusuPetrut** [comment]
Now it doesn't stop , just like reading hardware state for each channel all the time : 
![capture](https://user-images.githubusercontent.com/17450475/52670441-36b88780-2f21-11e9-8127-3641697601d6.JPG)

**Comment by RusuPetrut** [comment]
In the previous `git pull` i power on all the channels , after the systems starts it automatically stop each channel one by one

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52670926-7fbd0b80-2f22-11e9-98d7-4f067341c1d1.JPG)

**Comment by theyosh** [comment]
Hmm, about changing the state I have no clue at the moment. But I add some debug code. As I do not know sure if it does repeat that much.

So could you try to count how often the text is passing by: 'Meross switch init'

`git pull` and restart

There is also a lot of debug logging from the meross library

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52671089-d9253a80-2f22-11e9-9917-b401eb826cea.JPG)

**Comment by theyosh** [comment]
Ok, but as far I can see now, it is working. Only the changing the power switch does give an error.

But, for now, can you see the power switches on the web interface? And it is getting late here, so I have to stop for now. So I hope that you at least 4 power switches in the web interface right now

**Comment by RusuPetrut** [comment]
Add new channel: 1
Meross switch init
(<meross_iot.supported_devices.power_plugs.Mss425e object at 0x6b9b3ab0>, 1)
Add new channel: 2
Meross switch init
(<meross_iot.supported_devices.power_plugs.Mss425e object at 0x6b9b3ab0>, 2)
Add new channel: 3
Meross switch init
(<meross_iot.supported_devices.power_plugs.Mss425e object at 0x6b9b3ab0>, 3)
Add new channel: 4
Meross switch init
(<meross_iot.supported_devices.power_plugs.Mss425e object at 0x6b9b3ab0>, 4)
DONE scanning

**Comment by RusuPetrut** [comment]
Yes of course , thank you for for your support ! Have a good night !

**Comment by theyosh** [comment]
Ah thanks for the update... this looks pretty cool. So you have the power switches on the web interface. 

Then we have to see what happens when you try to toggle a power switch. Just toggle one channel just once.... When I have time I can see what I can do

**Comment by RusuPetrut** [comment]
All switches work on web interface. 💪

**Comment by theyosh** [comment]
Whaaat!!! You can toggle them!!!! NICE!!!! Party time...!

**Comment by RusuPetrut** [comment]
Great work !

**Comment by theyosh** [comment]
So, after two days, is it still working?

**Comment by RusuPetrut** [comment]
Hi , yesterday i've done a `git pull ` and now i receive this error : 

![capture](https://user-images.githubusercontent.com/17450475/52810269-d8171900-309a-11e9-8b90-3ed27c7c03d3.JPG)

**Comment by theyosh** [comment]
Is it started with python 2 by any chance?  And you are still on branch issue/254?

**Comment by RusuPetrut** [comment]
Yes , i am in branch issue/254 and python 3

**Comment by RusuPetrut** [comment]
Ok , now for no reason is working again 👍

**Comment by theyosh** [comment]
Hmm, I have also changed some debug code for the excessive logging of the Meross module. Is there now less verbose debug in the screen? 

But that error is a bit strange. Could maybe triggered from somewhere else...

**Comment by theyosh** [comment]
Did you already saved the power switches settings such as a name and power usage info? This can maybe interfere somehow with starting up

**Comment by theyosh** [comment]
Hi, could you do one more test for me? I did optimize the code a bit. So if you can do a final `git pull` on branch `issue/254` and restart the software. I expect no problems. So if you can confirm that, I will merge this with the master branch so that is available for everybody.

**Comment by RusuPetrut** [comment]
Hi , sorry for this late reply , i was going to write today with test results . So , i added one more power plug in Meross App and everything stop working . I've done `git pull ` now but the system restart over and over itself 

![capture](https://user-images.githubusercontent.com/17450475/52902378-6c5aba80-3218-11e9-8434-df7fb95da68c.JPG)

**Comment by theyosh** [comment]
Euh, I do not understand 'Adding more power plug to Meross App'. I just readout your hole device, and therefore you should have 4 power switches in my software. If it is possible to delete power plugs and therefore channels on your App, I do not know what will happen.

But for now, I think you have to manually change the settings.cfg file, and delete all references of your Meross power switches. So that only your meross cloud login details are available. Then restart the software and see what happens. Do not change the settings through the web interface.

It is hard to test without a good starting point, as I do not have a clue. But as far I can judge, it should still be able to at least login to the cloud and readout the power plugs.

**Comment by RusuPetrut** [comment]
Ok , my mistake , sorry  ! I added a power plug next to previous one and because they look the same i remove the old plug that worked with Terrarium Pi . Now everything works great !

**Comment by RusuPetrut** [comment]
Now comes the question , how can i add a second Meross power plug to Terrarium Pi . Is this possible ? Sorry for my insistence

**Comment by theyosh** [comment]
Euh, your terminology is confusing. You have now 4 power switches in TerrariumPI. Is this correct? Please answer this.

So, then I assume that you mean that you want to add a second device with multiple channels. Is this correct? If so, then you have to add that device to your Meross cloud. I do not know how this works from you phone. But as soon as you have added a second device, you should restart terrariumpi. Your Meross switches can only 'be added' with scanning during the startup of my software. When this is done you should have 8 power switches, with two times channel X names.

Else you have to enter the username and password for every powerswitch in my system.

**Comment by RusuPetrut** [comment]
Yes , how i have 4 power switches that appear in TerrariumPi .

 1 Power Plug = 4 power switches 

I've done browser cache cleaning and now all 8 power switches appear in TerrariumPi . Sorry about my confusing terminology.  

If you need further testing let me know . From my point of view , everything is working well now

**Comment by theyosh** [comment]
Have you already tried to change the name of the power switches in my interface? So that you can add your own name per switch? And when that does work and your switches are showing your entered names, try a restart. That would the last test I think before I can call it stable.

**Comment by RusuPetrut** [comment]
Ok , changed the names of the power switches in my interface . After restart , all settings are kept by the system .

**Comment by theyosh** [comment]
And no startup errors...? nice!

**Comment by RusuPetrut** [comment]
Yes , no startup errors

**Comment by RusuPetrut** [comment]
![capture](https://user-images.githubusercontent.com/17450475/52905043-d6855680-323c-11e9-924c-56ecf4f85d26.JPG)

**Comment by theyosh** [comment]
Hi,

I made an code error regarding the variable names for this power switch. Therefore I had to rename all values. It has been merged in the master branch.

So for you right now, do the following:

1. `git checkout master`
2. `git pull`
3. `nano settings.cfg`

And rename 'merros' to 'meross'  (3 times) and save the settings.cfg file. Then restart the software. That should do it for you for going back to the master branch.
